ocaml/xapi-storage-script/main.ml

```diff
@@ -814,924 +814,1003 @@ let choose_datapath ?(persistent = true) domain response =
-  let vdi_data_destroy_impl dbg sr vdi =
-    wrap
-    @@
-    let* sr = Attached_SRs.find sr in
-    let vdi = Storage_interface.Vdi.string_of vdi in
-    let* () =
+
+    let vdi_disable_cbt_impl dbg sr vdi =
+      wrap
+      @@
+      let* sr = Attached_SRs.find sr in
+      let vdi = Storage_interface.Vdi.string_of vdi in
       return_volume_rpc (fun () ->
-          Volume_client.data_destroy (volume_rpc ~dbg) dbg sr vdi
+          Volume_client.disable_cbt (volume_rpc ~dbg) dbg sr vdi
       )
-    in
-    set ~dbg ~sr ~vdi ~key:_vdi_type_key ~value:"cbt_metadata"
-  in
-  S.VDI.data_destroy vdi_data_destroy_impl ;
+
+    let vdi_list_changed_blocks_impl dbg sr vdi vdi' =
+      wrap
+      @@
+      let* sr = Attached_SRs.find sr in
+      let vdi, vdi' = Storage_interface.Vdi.(string_of vdi, string_of vdi') in
+      let ( let* ) = Lwt.bind in
+      let* result =
+        return_volume_rpc (fun () ->
+            (* Negative lengths indicate that we want the full length. *)
+            Volume_client.list_changed_blocks (volume_rpc ~dbg) dbg sr vdi vdi'
+              0L (-1)
+        )
+      in
+      let proj_bitmap r = r.Xapi_storage.Control.bitmap in
+      Lwt.return (Result.map proj_bitmap result)
+
+    let vdi_data_destroy_impl dbg sr vdi =
+      wrap
+      @@
+      let* sr = Attached_SRs.find sr in
+      let vdi = Storage_interface.Vdi.string_of vdi in
+      let* () =
+        return_volume_rpc (fun () ->
+            Volume_client.data_destroy (volume_rpc ~dbg) dbg sr vdi
+        )
+      in
+      set ~dbg ~sr ~vdi ~key:_vdi_type_key ~value:"cbt_metadata"
+  end
+
+module DPImpl =
+functor
+  (M : META)
+  ->
+  struct
+    module VDI = VDIImpl (M)
+
+    let dp_destroy2 dbg dp sr vdi' vm' _allow_leak =
+      (let vdi = Storage_interface.Vdi.string_of vdi' in
+       let domain = domain_of ~dp ~vm' in
+       Attached_SRs.find sr >>>= fun sr ->
+       (* Discover the URIs using Volume.stat *)
+       VDI.stat ~dbg ~sr ~vdi >>>= fun response ->
+       ( match
+           List.assoc_opt _clone_on_boot_key response.Xapi_storage.Control.keys
+         with
+       | None ->
+           return response
+       | Some temporary ->
+           VDI.stat ~dbg ~sr ~vdi:temporary
+       )
+       >>>= fun response ->
+       choose_datapath domain response >>>= fun (rpc, _datapath, uri, domain) ->
+       return_data_rpc (fun () ->
+           Datapath_client.deactivate (rpc ~dbg) dbg uri domain
+       )
+       >>>= fun () ->
+       return_data_rpc (fun () ->
+           Datapath_client.detach (rpc ~dbg) dbg uri domain
+       )
+      )
+      |> wrap
+  end
+
+(* Bind the implementations *)
+let bind ~volume_script_dir =
+  (* Each plugin has its own version, see the call to listen
+     where `process` is partially applied. *)
+  let module S = Storage_interface.StorageAPI (Rpc_lwt.GenServer ()) in
+  let module RuntimeMeta = struct
+    let volume_script_dir = volume_script_dir
+
+    let version = version
```

Comment:
If I understood correctly, version is shared among all the modules that implement storage functions. Why is it needed in RuntimeMeta? It looks like it depends on the global variable anyway, so why bother with the indirection?